### DISCUSSION TITLE
x-axis customizable

### DISCUSSION THREAD

**Discussion by carmelolg** [patrykandpatrick/vico#497] [Questions] [answered]
Can you explain me how can I customize the x-axis with string value? 
Following the [doc](https://patrykandpatrick.com/vico/wiki/axes) I'm not able to set the _horizontalAxisValueFormatter_ on my chart. 

I'm not using the "Composable" annotation and I'm using directly the object declared on the fragment layout.
Are there some SDK's API for setting the y-axis formatter directly by XML object param?

Just to be clear, I'm using the library as the following:

```
val p = ChartEntryModelProducer()
val entry: ChartEntry = entryOf(10, 1)
val entry2: ChartEntry = entryOf(20, 2)
val entry3: ChartEntry = entryOf(30, 3)
val entry4: ChartEntry = entryOf(40, 1)
val entry5: ChartEntry = entryOf(50, 2)
p.setEntries(listOf(entry, entry2, entry3, entry4, entry5))
binding.chartView.setModel(p.getModel())
```

Could be useful if I could put directly a String on the ChartEntry instance like 
```
val entry: ChartEntry = entryOf("Custom Label", 1)
```

**Comment by patrickmichalik** [comment, marked as answer]
Hello! As I understand, you’re referring to _x_-axes. For `horizontalAxisValueFormatter`, depending on the position of your `HorizontalAxis`, please replace `AxisPosition.Horizontal` with `AxisPosition.Horizontal.Top` or `AxisPosition.Horizontal.Bottom`. You can then apply the `AxisValueFormatter` as [here](https://github.com/patrykandpatrick/vico/blob/v1.13.1/sample/src/main/java/com/patrykandpatrick/vico/sample/showcase/charts/Chart6.kt#L93).

Regarding storing labels in `ChartEntry` instances, because chart data can be handled asynchronously, this isn’t entirely viable. A synchronous label update may be required in the event of a locale change, for example. Also, since inferred _x_ coordinates provide far less flexibility than explicit ones, two systems would have to exist in parallel, which would introduce a significant amount of complexity. Another issue is that such a solution wouldn’t necessarily be compatible with some of Vico’s more advanced APIs, such as `AxisItemPlacer`. Ultimately, charts are numerical, so a numerical approach provides the greatest flexibility and predictability. The difference between specifying labels at the `ChartEntry` level and using an `AxisValueFormatter` that reads from a map linking _x_ values to labels is rather minor—typically a few lines of code—but the latter approach is more suitable for asynchronous data updates and more adaptable.

**Reply by carmelolg** [reply]
Thanks for explanation, I understood.
Now it works.
C.

**Reply by patrickmichalik** [reply]
I’m glad to hear that. You’re welcome!
